### PR TITLE
Implement CRUD for serving-endpoints

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/10425?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10425/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10425
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Implement CRUS for endpoints.

TODOs:

- [x] Add tests

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:75]: We have a list of retry codes somewhere else that we use for tracking, right? Can we refer to that same list here

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:65]: `posixpath.append()`?

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:75]: can we use the default (_TRANSIENT_FAILURE_RESPONSE_CODES)?

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [inline mlflow/deployments/databricks/__init__.py:98]: will add docstrings later in a follow-up pr

File: mlflow/deployments/__init__.py
Comment (harupy) [inline mlflow/deployments/__init__.py:19]: This is for docs.

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:65]: Can we support `get_deploy_client("databricks://<scope>:<prefix>")`?

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:101]: @sueann To update rate limits, are users expected to supply the whole config?

Comment (sueann) [reply mlflow/deployments/databricks/__init__.py:101]: yeah. we could have made it a PATCH update but i think the decision to make it a PUT was because the `key` is not well-defined (??). @mingyu89?

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:68]: This timeout is too low for predict - can we use the same timeout used by gateway client? It can take O(minutes) to compute completions on a large amounts of text

For CRUD, I think we should retry with exponential backoff with the same # of retries & same timeout as we use for gateway (I think it's more than a 10 second timeout)

Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:68]: gateway uses 300. I'll use it.

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py]: replaced `"databricks"` with `self.target_uri`.

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [inline mlflow/deployments/databricks/__init__.py:94]: We need to support `tags`?

https://docs.databricks.com/api/workspace/servingendpoints/create

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [inline mlflow/deployments/databricks/__init__.py:68]: gateways uses an environment variable for `timeout`. not sure if we need to do the same thing here.

Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:68]: Not an environment variable, just a constant variable.

Comment (dbczumar) [reply mlflow/deployments/databricks/__init__.py:68]: yeah, we should rename the var and do the same

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [reply mlflow/deployments/databricks/__init__.py:94]: Great catch - yes :)

Comment (harupy) [reply mlflow/deployments/databricks/__init__.py]: updated

File: mlflow/deployments/databricks/constants.py
Comment (BenWilson2) [inline mlflow/deployments/databricks/constants.py:2]: I'm of the opinion of having an environment variable that uses this constant as a default might be worthwhile just to be on the safe side and give us some flexibility in controlling the behavior in case a user deploys something complex with a very high inference latency. WDYT?

File: mlflow/deployments/databricks/__init__.py
Comment (mingyu89) [reply mlflow/deployments/databricks/__init__.py:107]: Yes. Here's example of the payload data
```
data = {
    "rate_limits": [{"key": "user", "renewal_period": "minute", "calls": 5}]
}
```
If we do patch, we have to compare all three fields to do the remove/update.
And currently the max number of limits people can specify is 2 - one user limit and one endpoint limit.

Comment (mingyu89) [reply mlflow/deployments/databricks/__init__.py:107]: Rate limit PUT has its own request.
Screenshot 2023-11-16 at 12 31 49 PM

File: mlflow/deployments/databricks/constants.py
Comment (dbczumar) [reply mlflow/deployments/databricks/constants.py:2]: @harupy I agree with @BenWilson2 . Can we also call this `MLFLOW_DEPLOYMENT_PREDICT_TIMEOUT` and use it for Databricks / MLflow deployment clients? I also think 300 seconds is too long for CRUD but correct for predict(). Can we only apply this timeout to predict?

Comment (dbczumar) [reply mlflow/deployments/databricks/constants.py:2]: @harupy For CRUD, can we use the same timeout we use for MLflow Tracking?

File: examples/deployments/databricks.py
Comment (harupy) [inline examples/deployments/databricks.py:46]: `create_endopoint` worked yesterday, but doesn't today. Did we update something? @mingyu89 

```
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://e2-dogfood.staging.cloud.databricks.com/api/2.0/serving-endpoints/. Response text: {"error_code":"INVALID_PARAMETER_VALUE","message":"Please provide a valid endpoint task type."}
```

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:91]: @harupy feel free to do this in a follow-up PR, but I think it would be nice if we returned a `DatabricksEndpoint` object that wraps the response dictionary and exposes some properties nicely like `name` and endpoint type

File: mlflow/deployments/databricks/constants.py
Comment (harupy) [reply mlflow/deployments/databricks/constants.py]: updated

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:95]: Added a TODO comment.

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [inline mlflow/deployments/databricks/__init__.py:96]: @dbczumar We could a return dict-like object here:

```
class AttrDict(dict):
    """
    Examples
    --------
    >>> d = AttrDict({'a': 1, 'b': 2})
    >>> d.a
    1
    >>> d = AttrDict({'a': 1, 'b': {'c': 3, 'd': 4}})
    >>> d.b.c
    3
    """

    def __getattr__(self, attr):
        value = self[attr]
        if isinstance(value, dict):
            return AttrDict(value)
        return value
```

File: examples/deployments/databricks.py
Comment (harupy) [reply examples/deployments/databricks.py:46]: or I'm just missing something?

File: mlflow/deployments/databricks/__init__.py
Comment (serena-ruan) [inline mlflow/deployments/databricks/__init__.py:111]: Check with Mingyu and sync with databricks APIs

File: mlflow/deployments/databricks/__init__.py
Comment (serena-ruan) [reply mlflow/deployments/databricks/__init__.py:96]: Agreed from Corey!

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py]: @mingyu89 Thanks! Is this a new endpoint we added for setting rate limits?

Comment (mingyu89) [reply mlflow/deployments/databricks/__init__.py]: It's a new API for updating(PUT) rate limits. 
Rate limits can also be specified during endpoint creation.
The rate limits can only be set on foundation models and external models.

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:112]: @mingyu89 Let me know if there is a better way to support rate limits :)

File: examples/deployments/databricks.py
Comment (mingyu89) [reply examples/deployments/databricks.py:46]: There's a new API update got merged yesterday. This notebooks is the latest state:
https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/4169768892691893/command/4169768892691894

Basically, the "task" is moved from a top level parameter of the request to the external_model object.
And the `openai_api_key` field only takes databricks secrets with format `{{secrets/<scope>/<key>}}`. It no longer takes plain text API token.

File: mlflow/deployments/databricks/__init__.py
Comment (mingyu89) [reply mlflow/deployments/databricks/__init__.py]: I updated the notebook to include creating endpoint with rate limits and set rate limits.
Setting rate limit with empty array is to delete the rate limits. https://e2-dogfood.staging.cloud.databricks.com/?o=6051921418418893#notebook/4169768892691893/command/4169768892691894

File: examples/deployments/databricks.py
Comment (harupy) [reply examples/deployments/databricks.py:46]: got it, updating the code

Comment (harupy) [reply examples/deployments/databricks.py]: Updated. The example is working fine now.

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py]: @mingyu89 Is this a valid request?

```
url = (
    host + "api/2.0/serving-endpoints/mingyu_external_openai_embeddings_endpoint/config"
)

data = {
    "name": "mingyu_openai_embeddings_endpoint",
    "served_entities": [
        {
            "name": "updated-served-entity-name",
            "external_model": {
                "name": "text-embedding-ada-002",
                "provider": "openai",
                "task": "llm/v1/embeddings",
                "openai_config": {
                    "openai_api_key": "{{secrets/mingyu-external-model/openai}}"
                },
            },
        }
    ],
    "rate_limits": [
        {
            "key": "user",
            "renewal_period": "minute",
            "calls": 3,
        }
    ],
}

response = requests.put(url, headers=headers, json=data)
```

I got this error:

```
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://e2-dogfood.staging.cloud.databricks.com/api/2.0/serving-endpoints/test-endpoint-816e1f87-2deb-4e09-a69e-0656703932f9/config. Response text: {"error_code":"BAD_REQUEST","message":"Config needs to be set and config.served_models must contain at least one element"}
```

Comment (harupy) [reply mlflow/deployments/databricks/__init__.py]: Never mind. I found a mistake.

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:85]: Can we call this `DatabricksEndpoint`?

Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:85]: When returning from `create_endpoint`, `get_endpoint`, etc, can we call this `DatabricksEndpoint`? Looks nicer for users than `AttrDict`

File: mlflow/deployments/constants.py
Comment (dbczumar) [inline mlflow/deployments/constants.py:6]: Should we move this to `environment_variables.py`?

File: mlflow/deployments/databricks/__init__.py
Comment (dbczumar) [inline mlflow/deployments/databricks/__init__.py:129]: Can we extract the `endpoints` field out of the response and return just that part?

**Review by dbczumar — APPROVED** [review]
LGTM with a few small comments. Thanks @harupy  - tried it out, works great! Go team!

File: mlflow/utils/__init__.py
Comment (BenWilson2) [inline mlflow/utils/__init__.py:273]: Perhaps we could subclass the deployment implementation from this generic abstract to get the specific Databricks naming convention? This could definitely be used elsewhere!

**Review by BenWilson2 — APPROVED** [review]
LGTM after pending comments are addressed. Great progress on this @harupy 👍

File: mlflow/deployments/constants.py
Comment (harupy) [reply mlflow/deployments/constants.py:6]: I defined this here to avoid conflicts. After merging this PR, can you request a change to move this to `environment_variables.py`?

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:129]: Great idea!

File: mlflow/deployments/databricks/__init__.py
Comment (harupy) [reply mlflow/deployments/databricks/__init__.py:85]: Updated, users probably won't see the class name tho.

File: mlflow/utils/__init__.py
Comment (harupy) [reply mlflow/utils/__init__.py:273]: updated!

**Pull request merged**
